File: app/src/main/java/knf/kuma/animeinfo/ActivityImgFull.java
package knf.kuma.animeinfo;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.ParcelFileDescriptor;
import android.provider.MediaStore;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.PopupMenu;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.squareup.picasso.Callback;

import java.io.FileOutputStream;

import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.R;
import knf.kuma.commons.PicassoSingle;
import xdroid.toaster.Toaster;

/**
 * Created by Jordy on 15/01/2018.
 */

public class ActivityImgFull extends AppCompatActivity implements PopupMenu.OnMenuItemClickListener {

    @BindView(R.id.img)
    ImageView imageView;
    @BindView(R.id.anchor)
    View anchor;
    private Bitmap bitmap;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.layout_img_big);
        ButterKnife.bind(this);
        PicassoSingle.get(this).load(getIntent().getData()).into(imageView, new Callback() {
            @Override
            public void onSuccess() {
                bitmap=((BitmapDrawable) imageView.getDrawable()).getBitmap();
            }

            @Override
            public void onError() {

            }
        });
        imageView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                PopupMenu popupMenu = new PopupMenu(ActivityImgFull.this, anchor);
                popupMenu.inflate(R.menu.menu_img);
                popupMenu.setOnMenuItemClickListener(ActivityImgFull.this);
                popupMenu.show();
                return true;
            }
        });
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.download:
                Intent i = new Intent(Intent.ACTION_CREATE_DOCUMENT)
                        .addCategory(Intent.CATEGORY_OPENABLE)
                        .setType("image/jpg")
                        .putExtra(Intent.EXTRA_TITLE, getIntent().getStringExtra("title")+".jpg");
                startActivityForResult(i, 556);
                break;
            case R.id.share:
                Intent intent = new Intent(Intent.ACTION_SEND)
                        .setType("image/*")
                        .putExtra(Intent.EXTRA_SUBJECT, getIntent().getStringExtra("title"))
                        .putExtra(Intent.EXTRA_STREAM, Uri.parse(MediaStore.Images.Media.insertImage(getContentResolver(),bitmap , "", "")));
                startActivity(Intent.createChooser(intent, "Compartir..."));
                break;
        }
        return true;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, final Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        final MaterialDialog dialog = new MaterialDialog.Builder(this)
                .progress(true, 100)
                .content("Guardando...")
                .cancelable(false)
                .build();
        dialog.show();
        AsyncTask.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    ParcelFileDescriptor pfd = getContentResolver().
                            openFileDescriptor(data.getData(), "w");
                    FileOutputStream fileOutputStream =
                            new FileOutputStream(pfd.getFileDescriptor());
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 100, fileOutputStream);
                    fileOutputStream.close();
                    pfd.close();
                    Toaster.toast("Image guardada!");
                } catch (Exception e) {
                    e.printStackTrace();
                    Toaster.toast("Error al guardar imagen");
                }
                dialog.dismiss();
            }
        });
    }

    @Override
    public void onBackPressed() {
        supportFinishAfterTransition();
    }
}

File: app/src/main/java/knf/kuma/animeinfo/ActivityImgFull.kt
package knf.kuma.animeinfo

import android.content.Intent
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.net.Uri
import android.os.Bundle
import android.preference.PreferenceManager
import android.provider.MediaStore
import android.view.MenuItem
import android.widget.ProgressBar
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.PopupMenu
import com.google.android.material.snackbar.Snackbar
import com.squareup.picasso.Callback
import knf.kuma.R
import knf.kuma.commons.*
import kotlinx.android.synthetic.main.layout_img_big.*
import okhttp3.Request
import org.jetbrains.anko.doAsync
import org.json.JSONObject
import xdroid.toaster.Toaster
import java.io.FileOutputStream
import java.net.URLEncoder

class ActivityImgFull : AppCompatActivity(), PopupMenu.OnMenuItemClickListener {

    private var bitmap: Bitmap? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.layout_img_big)
        PicassoSingle[this].load(intent.data).into(img, object : Callback {
            override fun onSuccess() {
                bitmap = (img.drawable as BitmapDrawable).bitmap
            }

            override fun onError() {
                Toaster.toast("Error al cargar imagen")
                finish()
            }
        })
        img.setOnLongClickListener {
            if (bitmap != null) {
                val popupMenu = PopupMenu(this@ActivityImgFull, anchor!!)
                popupMenu.inflate(R.menu.menu_img)
                popupMenu.setOnMenuItemClickListener(this@ActivityImgFull)
                popupMenu.show()
            }
            true
        }
        searchInMAL()
    }

    private fun searchInMAL() {
        if (PreferenceManager.getDefaultSharedPreferences(this).getBoolean("scale_img", true))
            doAsync {
                val snackbar = img.showSnackbar("Buscando mejor resolucion...", Snackbar.LENGTH_INDEFINITE)
                try {
                    val title = intent.getStringExtra("title")
                    val response = Request.Builder()
                            .url("https://api.jikan.moe/v3/search/anime?q=${URLEncoder.encode(title, "utf-8")}&page=1")
                            .build().execute()
                    if (response.code() != 200)
                        throw IllegalStateException("Response code: ${response.code()}")
                    val results = JSONObject(response.body()!!.string()).getJSONArray("results")
                    response.close()
                    for (i in 0..(results.length() - 1)) {
                        val json = results.getJSONObject(i)
                        val name = json.getString("title").toLowerCase()
                        if (title.toLowerCase() == name) {
                            var imgUrl: String = json.getString("image_url")
                            try {
                                val picturesResponse = Request.Builder().url("https://api.jikan.moe/v3/anime/${json.getString("mal_id")}/pictures").build().execute()
                                if (picturesResponse.code() != 200)
                                    throw IllegalStateException("Response code: ${picturesResponse.code()}")
                                val picturesArray = JSONObject(picturesResponse.body()!!.string()).getJSONArray("pictures")
                                picturesResponse.close()
                                imgUrl = picturesArray.getJSONObject(picturesArray.length() - 1).getString("large")
                            } catch (e: Exception) {
                                //
                            }
                            doOnUI {
                                PicassoSingle[this@ActivityImgFull].load(imgUrl).into(img, object : Callback {
                                    override fun onSuccess() {
                                        bitmap = (img.drawable as BitmapDrawable).bitmap
                                    }

                                    override fun onError() {

                                    }
                                })
                            }
                            break
                        }
                    }
                    snackbar.safeDismiss()
                } catch (e: Exception) {
                    e.printStackTrace()
                    snackbar.safeDismiss()
                }
            }
    }

    override fun onMenuItemClick(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.download -> {
                val i = Intent(Intent.ACTION_CREATE_DOCUMENT)
                        .addCategory(Intent.CATEGORY_OPENABLE)
                        .setType("image/jpg")
                        .putExtra(Intent.EXTRA_TITLE, intent.getStringExtra("title") + ".jpg")
                startActivityForResult(i, 556)
            }
            R.id.share -> try {
                val intent = Intent(Intent.ACTION_SEND)
                        .setType("image/*")
                        .putExtra(Intent.EXTRA_TEXT, intent.getStringExtra("title"))
                        .putExtra(Intent.EXTRA_STREAM, Uri.parse(MediaStore.Images.Media.insertImage(contentResolver, bitmap, "", "")))
                startActivity(Intent.createChooser(intent, "Compartir..."))
            } catch (e: Exception) {
                Toaster.toast("Error al compartir")
            }
        }
        return true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        val snackbar = img.createSnackbar("Guardando...", Snackbar.LENGTH_INDEFINITE)
        val progressBar = ProgressBar(this).also {
            it.isIndeterminate = true
        }
        (snackbar.view as Snackbar.SnackbarLayout).addView(progressBar)
        snackbar.show()
        doAsync {
            try {
                val pfd = contentResolver.openFileDescriptor(data!!.data!!, "w")
                val fileOutputStream = FileOutputStream(pfd!!.fileDescriptor)
                bitmap!!.compress(Bitmap.CompressFormat.JPEG, 100, fileOutputStream)
                fileOutputStream.close()
                pfd.close()
                snackbar.safeDismiss()
                img.showSnackbar("Image guardada!")
            } catch (e: Exception) {
                e.printStackTrace()
                snackbar.safeDismiss()
                img.showSnackbar("Error al guardar imagen")
            }
        }
    }

    override fun onBackPressed() {
        supportFinishAfterTransition()
    }
}
